feat: Batch job - Spellcheck

jeremyarancio opened the pull request:
# Objective of the PR:
A new model based on Mistral-7B LLM has been developed to tackle the OCR-parsed ingredients spellcheck problem (see [#spellcheck](https://openfoodfacts.slack.com/archives/CC5B3JZ0U)). While this solution doesn't fit Open Food Facts' production infrastructure, we've created a GPU-powered batch inference pipeline on GCP to process thousands of products.

See demo: [Ingredient Spellcheck](https://huggingface.co/spaces/jeremyarancio/ingredients-spellcheck)

### What
- Set up Batch Job with GCP for Robotoff Tasks.
- We start with Ingredients-Spellcheck


### Part of 
- Spellcheck work ([link](https://docs.google.com/document/d/14dUeVTxJVMIdognHerVL4cuFKcj5hPsu8ZT5-3Js23I/edit))
- Discussion Robotoff ([link](https://docs.google.com/document/d/1nTaHJOKC5-A9-qRUGBl9sqk34q3bg-Sft4E6ePwwXKc/edit#heading=h.f10dlj3eczcf))

jeremyarancio commented:
Rest to do:
* Transfert batch launch from api endpoint to CLI
* Set up API security key for Batch Data import, triggered by GCP batch job once finished
* Improve InsightsImport to consider several use cases (see [Robotoff doc](https://docs.google.com/document/d/1RUfmWHjtFVaBcvQ17YfXu6FW6oRFWg-2lncljG0giKI/edit#heading=h.s3zhmssf4u79))

raphael0202 commented on robotoff/insights/importer.py:
```diff
@@ -1475,6 +1475,35 @@ def compute_crop_bounding_box(
+        predictions: list[Prediction],
+        product_id: ProductIdentifier,
+    ) -> Iterator[ProductInsight]:
+        # Only one prediction
```
Possibly more than 1, no?

jeremyarancio replied:
There's only one correction, no?
Or maybe I didn't understand what it means

jeremyarancio commented on pyproject.toml:
```diff
@@ -58,7 +58,7 @@ jsonschema = "~4.4.0"
 Pillow = ">=9.3,<10.4"
 numpy = "~1.26.4"
-protobuf = "~3.19.0"
+protobuf = "^3.19.0"
```
Google-cloud-storage needed a more recent version of protobuf
I increased it to 3.20.3

jeremyarancio commented on robotoff/batch/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    LOGGER.info(f"Batch import results: {repr(import_results)}.")
+
+
+def _generate_predictions_from_batch(
```
I agree
I modified the code and remove the overall abstraction of the code

jeremyarancio commented on robotoff/cli/main.py:
```diff
@@ -998,5 +998,22 @@ def create_migration(
+    from robotoff.utils import get_logger
+    from robotoff.types import BatchJobType
+
+    if job_type not in BatchJobType.__members__:
```
Nope since job_type is a string and not an enum

```python
from enum import Enum
class Model(Enum):
    a = 1
    b = 2
try: 
    "a" in Model
except Exception as e:
    print(e)

print("a" in Model.__members__) 
```

```
unsupported operand type(s) for 'in': 'str' and 'EnumType'
True
<ipython-input-2-32a1e4cd847a>:6: DeprecationWarning: in 3.12 __contains__ will no longer raise TypeError, but will return True or
False depending on whether the value is a member or the value of a member
  "a" in Model
```

raphael0202 replied:
Ah yes, usually I do a BatchJobType[job_type] and catch KeyError, but this approach is fine too :+1:

jeremyarancio commented on robotoff/batch/__init__.py:
```diff
@@ -0,0 +1,105 @@
+                barcode=row[features_dict["barcode"]],
+                predictor_version=PREDICTOR_VERSION,
+                predictor=PREDICTOR,
+            )
```
I'm afraid that when I query the off database, I got limited by the number of queries

raphael0202 replied:
My bad, actually the ingredient text is not linked to the `source_image`, as we don't know which image it comes from.

jeremyarancio commented:
Black fails on `os.getenv("BATCH_JOB_KEY")}`, which is normal since it is not added to Github secret
This job key is used by the import_batch endpoint to secure it. Only the batch job docker image owns it.
We will have to add a key to Github secrets for the API, but also during the deployment of the image in GCP (CI/CD to make things easier?)

jeremyarancio commented:
Quick note about error handling:
Falcon doesn't seem to handle custom exception error (such `APITokenError(Exception)`) and return status 500 instead by default

jeremyarancio commented:
I implemented the following changes:
* Add make deploy-spellcheck in the Makefile
* Modified the query to extract batch based on `popularity_key`
* Delve into Google credentials to remove the `credentilal.json` folder necessity. We'll use a [service_account](https://cloud.google.com/iam/docs/service-account-overview) instead (**Need to generate it for production with the correct roles: incoming**)
* Added environment key to the batch job 
* Update Importer (*still a bug, needs to solve it*)
* The overall process is tested, and everything looks fine. (Process 10000 products in 20 minutes)

jeremyarancio commented:
Cannot explain why test fail with DeepSource, and not for other identical cases

![image](https://github.com/user-attachments/assets/e516ad15-a36e-45d4-ae5a-cfb0f6dc8cf4)

raphael0202 commented:
Deepsource generates a lot of false positive. I thought about disabling it to be honest.

jeremyarancio commented:
Improvement ideas:
* Generate an unique id for each data processed in the bucket (reason: enable debugging)
